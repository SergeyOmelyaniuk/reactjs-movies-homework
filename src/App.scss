@import './scss/variables.scss';

.wrapper {
	color: white;
	font-family: 'Roboto', sans-serif;
	background-color: $colorBackground;
	min-height: 100%;
}

.container {
	max-width: 1585px;
	margin: 0 auto;
}
