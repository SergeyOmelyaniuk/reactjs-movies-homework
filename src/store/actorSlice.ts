import { createSlice, createAsyncThunk, PayloadAction } from '@reduxjs/toolkit';
import { Actor, MovieAPI, MovieImages } from '../types';

export const fetchActorProfile = createAsyncThunk(
	'actor/fetchActorProfile',
	async function (id: string, { rejectWithValue, getState }) {
		const state = getState() as any;
		const { languageSelected } = state.language;

		try {
			const [actor, images, movies] = await Promise.all([
				fetch(
					`https://api.themoviedb.org/3/person/${id}?api_key=${process.env.REACT_APP_API_KEY}&language=${languageSelected.value}`
				),
				fetch(
					`https://api.themoviedb.org/3/person/${id}/images?api_key=${process.env.REACT_APP_API_KEY}`
				),
				fetch(
					`https://api.themoviedb.org/3/person/${id}/combined_credits?api_key=${process.env.REACT_APP_API_KEY}&language=${languageSelected.value}`
				),
			]).then((data) => Promise.all(data.map((res) => res.json())));

			return { actor, images, movies };
		} catch (error) {
			return rejectWithValue(error.message);
		}
	}
);

export interface ActorState {
	status: string | null;
	actorProfile: Actor | null;
	actorImages: MovieImages[] | null;
	actorKnowBy: MovieAPI[] | null;
}

const initialState: ActorState = {
	status: null,
	actorProfile: null,
	actorImages: null,
	actorKnowBy: null,
};

const actorReducer = createSlice({
	name: 'actor',
	initialState,
	reducers: {},
	extraReducers: (builder) => {
		builder.addCase(fetchActorProfile.pending, (state) => {
			state.status = 'loading';
		});
		builder.addCase(
			fetchActorProfile.fulfilled,
			//TODO
			(state, { payload }: PayloadAction<any>) => {
				state.status = 'fulfilled';
				state.actorProfile = payload.actor;
				state.actorImages = payload.images.profiles;
				state.actorKnowBy = payload.movies.cast;
			}
		);
		builder.addCase(fetchActorProfile.rejected, (state) => {
			state.status = 'failed';
		});
	},
});

export default actorReducer.reducer;
