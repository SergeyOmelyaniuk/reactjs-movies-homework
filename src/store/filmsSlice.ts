import { createSlice, createAsyncThunk, PayloadAction } from '@reduxjs/toolkit';
import { translate } from '../constants';
import getGenresFilms from '../helpers/getGenresFilms';
import { Movie, Genre } from '../types';

export const fetchFilms = createAsyncThunk(
	'todos/fetchFilms',
	async function (_, { rejectWithValue, getState, dispatch }) {
		const { films } = getState() as any;

		if (films.listGenres.length === 0) {
			await dispatch(fetchGenres());
		}

		const state = getState() as any;
		const { currentPage, categoryValue, listGenres, searchValue } = state.films;
		const url = searchValue
			? `${process.env.REACT_APP_API_URL}/search/movie?query=${searchValue}&api_key=${process.env.REACT_APP_API_KEY}&language=en-US&page=${currentPage}`
			: `${process.env.REACT_APP_API_URL}/movie/${categoryValue}?api_key=${process.env.REACT_APP_API_KEY}&language=en-US&page=${currentPage}`;

		try {
			const response = await fetch(url);

			const data = await response.json();

			return {
				totalPages: data.total_pages,
				results: getGenresFilms(listGenres, data.results),
			};
		} catch (error) {
			return rejectWithValue(error.message);
		}
	}
);

export const fetchGenres = createAsyncThunk(
	'todos/fetchGenres',
	async function (_, { rejectWithValue }) {
		try {
			const response = await fetch(
				`${process.env.REACT_APP_API_URL}/genre/movie/list?api_key=${process.env.REACT_APP_API_KEY}&language=en-US`
			);

			const data = await response.json();

			return data;
		} catch (error) {
			return rejectWithValue(error.message);
		}
	}
);

export interface FilmsState {
	categoryValue: string;
	currentPage: number;
	totalPages: number;
	listfilms: Movie[];
	status: null | string;
	listGenres: Genre[];
	searchValue: string;
}

export const initialState: FilmsState = {
	categoryValue: translate.en.categories[0].value,
	currentPage: 1,
	totalPages: 0,
	listfilms: [],
	status: null,
	listGenres: [],
	searchValue: '',
};

const filmsReducer = createSlice({
	name: 'films',
	initialState,
	reducers: {
		changeCategory(state, action: PayloadAction<string>) {
			state.categoryValue = action.payload;
			state.currentPage = 1;
		},
		setCurrentPage(state, action: PayloadAction<number>) {
			state.currentPage = action.payload;
		},
		setSearcValue(state, action: PayloadAction<string>) {
			state.searchValue = action.payload;
			state.currentPage = 1;
		},
	},
	extraReducers: (builder) => {
		builder.addCase(fetchFilms.pending, (state) => {
			state.status = 'loading';
		});
		builder.addCase(
			fetchFilms.fulfilled,
			(
				state,
				{ payload }: PayloadAction<{ totalPages: number; results: Movie[] }>
			) => {
				state.listfilms = payload.results;
				state.totalPages = payload.totalPages;
			}
		);
		builder.addCase(fetchFilms.rejected, (state) => {
			state.status = 'failed';
		});

		builder.addCase(fetchGenres.pending, (state) => {
			state.status = 'loading';
		});
		builder.addCase(
			fetchGenres.fulfilled,
			(state, { payload }: PayloadAction<{ genres: Genre[] }>) => {
				state.listGenres = payload.genres;
			}
		);
		builder.addCase(fetchGenres.rejected, (state) => {
			state.status = 'failed';
		});
	},
});

export const { changeCategory, setCurrentPage, setSearcValue } =
	filmsReducer.actions;

export default filmsReducer.reducer;
