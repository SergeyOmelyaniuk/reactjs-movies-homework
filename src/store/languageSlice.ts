import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { languages } from '../constants';
import { Language } from '../types';

export interface LanguageState {
	languageSelected: Language;
}

const initialState: LanguageState = {
	languageSelected: languages[0],
};

const languageReducer = createSlice({
	name: 'language',
	initialState,
	reducers: {
		setCurrentlanguage(state, action: PayloadAction<Language>) {
			state.languageSelected = action.payload;
		},
	},
});

export const { setCurrentlanguage } = languageReducer.actions;

export default languageReducer.reducer;
