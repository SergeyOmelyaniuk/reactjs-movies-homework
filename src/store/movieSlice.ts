import { createSlice, createAsyncThunk, PayloadAction } from '@reduxjs/toolkit';
import { MovieAPI, Genre, MovieDetails, MovieImages, Actor } from '../types';

export const fetchMovieDetails = createAsyncThunk(
	'movie/fetchMovieDetails',
	async function (id: string, { rejectWithValue, getState }) {
		const state = getState() as any;
		const { languageSelected } = state.language;
		// const { currentIdFilm } = state.film;
		//TODO

		try {
			const [movie, images, recommendations, actors] = await Promise.all([
				fetch(
					`https://api.themoviedb.org/3/movie/${id}?api_key=${process.env.REACT_APP_API_KEY}&language=${languageSelected.value}`
				),
				fetch(
					`https://api.themoviedb.org/3/movie/${id}/images?api_key=${process.env.REACT_APP_API_KEY}`
				),
				fetch(
					`https://api.themoviedb.org/3/movie/${id}/recommendations?api_key=${process.env.REACT_APP_API_KEY}&language=${languageSelected.value}&page=1`
				),
				fetch(
					`https://api.themoviedb.org/3/movie/${id}/credits?api_key=${process.env.REACT_APP_API_KEY}&language=${languageSelected.value}`
				),
			]).then((data) => Promise.all(data.map((res) => res.json())));

			return { movie, images, recommendations, actors };
		} catch (error) {
			return rejectWithValue(error.message);
		}
	}
);

export interface MovieState {
	movieDetails: MovieDetails | null;
	movieImages: MovieImages[] | null;
	moviesRecommendations: MovieAPI[] | null;
	topCast: Actor[] | null;
	status: string | null;
	listGenres: Genre[];
	//TODO
	// currentIdFilm: number;
}

const initialState: MovieState = {
	movieDetails: null,
	movieImages: null,
	moviesRecommendations: null,
	topCast: null,
	status: null,
	listGenres: [],
	//TODO
	// currentIdFilm: 460465,
};

const movieReducer = createSlice({
	name: 'movies',
	initialState,
	reducers: {},
	extraReducers: (builder) => {
		builder.addCase(fetchMovieDetails.pending, (state) => {
			state.status = 'loading';
		});
		builder.addCase(
			fetchMovieDetails.fulfilled,
			//TODO
			(state, { payload }: PayloadAction<any>) => {
				state.movieDetails = payload.movie;
				state.movieImages = payload.images.backdrops;
				state.moviesRecommendations = payload.recommendations.results;
				state.topCast = payload.actors.cast;
				state.status = 'fulfilled';
			}
		);
		builder.addCase(fetchMovieDetails.rejected, (state) => {
			state.status = 'failed';
		});
	},
});
//TODO
// export const { } = movieReducer.actions;

export default movieReducer.reducer;
