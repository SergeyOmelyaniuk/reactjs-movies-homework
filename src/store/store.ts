import { configureStore } from '@reduxjs/toolkit';
import languageReducer from './languageSlice';
import moviesReducer from './moviesSlice';
import movieReducer from './movieSlice';

export const store = configureStore({
	reducer: {
		language: languageReducer,
		movies: moviesReducer,
		movie: movieReducer,
	},
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
