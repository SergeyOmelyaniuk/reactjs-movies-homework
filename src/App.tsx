import React from 'react';
import './App.scss';
import ActorProfile from './pages/ActorProfile';
import MovieDetails from './pages/MovieDetails';
import Main from './pages/Main';
import Header from './сomponents/Header';
import ErrorBoundary from './сomponents/ErrorBoundary';
import { Route, Switch } from 'react-router-dom';

function App() {
	return (
		<div className='wrapper'>
			<Header />
			<div className='container'>
				<ErrorBoundary>
					<Switch>
						<Route exact path='/' component={Main} />
						<Route path='/movies/:movieId' component={MovieDetails} />
						<Route path='/actors/:actorId' component={ActorProfile} />
					</Switch>
				</ErrorBoundary>
			</div>
		</div>
	);
}

export default App;
