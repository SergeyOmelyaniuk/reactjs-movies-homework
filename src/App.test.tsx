import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { render } from '@testing-library/react';
import App from './App';
import { Provider } from 'react-redux';
import { store } from './store/store';

jest.mock('./pages/Main', () => () => 'Main');

describe('App', () => {
	test('should render', () => {
		const { container } = render(
			<Provider store={store}>
				<BrowserRouter>
					<App />
				</BrowserRouter>
			</Provider>
		);

		expect(container).toMatchSnapshot();
	});
});
