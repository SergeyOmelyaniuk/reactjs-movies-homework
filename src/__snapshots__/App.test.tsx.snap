// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render 1`] = `
<div>
  <div
    class="wrapper"
  >
    <header
      class="header"
    >
      <div
        class="container"
      >
        <a
          href="/"
        >
          <h1
            class="title"
          >
            title
          </h1>
        </a>
        <form
          class="search"
        >
          <input
            class="input"
            placeholder="Movies, person, movie theaters"
            type="text"
            value=""
          />
          <button
            class="button"
            type="submit"
          />
        </form>
        <div
          class="languageBar"
        >
          <div
            class="menu"
          >
            en
          </div>
        </div>
      </div>
    </header>
    <div
      class="container"
    >
      Main
    </div>
  </div>
</div>
`;
